README.md
# bot_skeleton
skeleton for tweetbots written in Python.

Created by Andrew Michaud ahead of #NaBoMaMo 2017


CHANGELOG
3.2.5 (2019-03-11):
html unescaping
* BUGFIX
    * both birdsite and mastodon batch reply would give HTML-encoded strings to the caller.
    this runs html.unescape on the strings to turn things like &gt; back into >.

3.2.4 (2019-03-08):
dm manual fix
* BUGFIX
    * birdsite dm was entirely broken without me noticing,
    because birdsite changed the API and tweepy hasn't yet adapted.
    fixed with a manual block similar to how caption upload is done.
    * add another check when procesing error to record,
    in case error can't be json dumped.

* DEV
    * make _handle_caption_upload and _upload_caption "private" with the leading underscore.
    * force keyword-only args in handle_error.
    * have batch_reply handle target handle either with or without a leading "@".

3.2.3 (2019-03-08):
more error handling
* BUGFIX
    * batch reply won't crash on a duplicate message on the birdsite output.

3.2.2 (2019-03-08):
correct batch reply
* BUGFIX
    * Batch reply for both birdsite and mastodon was broken.
    * Fix birdsite reply to get the full message text correctly.
    * Fix mastodon reply to get the correct target account or fail out.

3.2.1 (2019-03-07):
small fix
* BUGFIX
    * botskeleton send passed the wrong argument in,
    causing all calls to fail if only position argument text was provided.

3.2.0 (2019-03-07):
batch reply functionality
* FEATURES
    * Can now reply in batch to a target user with perform_batch_reply().
    * You need to supply a callback,
    and a target per service,
    and optionally some other fields,
    and the callback will be used to reply to a number of their recent tweets (number configurable)

* BUGFIXES
    * Permit writing lists as JSON from IterationRecord,
    needed to support lists of OutputRecord.
    * Make _repair a little more robust,
    and have it assume less.
    * Don't try to _repair legacy tweet records,
    this is the wrong behavior and I'm not sure what would happen.

* DEV
    * Make all API methods prefer keyword arguments,
    though positional arguments will still be allowed.
    * All internal methods now use only keyword arguments (or at least most do).
    * Use verbose docstrings that can be formatted as documentation,
    and eventually I'll do that.
    * Expect python3.7 support and update travisCI config appropriately.

3.1.3 (2019-03-06):
bugfix
* BUGFIX
    * _repair method would not handle extra keys correctly and would fail out.

3.1.2 (2019-03-05):
Dependency Updates
* DEV
    * Pin new dependencies.

3.1.1 (2019-03-05):
Dependency Updates
* DEV
    * Pin new dependencies.
    * Get fix from drewtilities.

3.1.0 (2018-09-23):
Various
* FEATURES
    * Can now upload captions when sending media posts.
    * Provide list of captions alongside files in API calls.
      If there are less captions than files, empty ("") captions will be filled in.

* BUGFIXES
    * Fix history corruption bug that led to nested "output_records".
    * Fix trying inactive outputs on media calls.

* DEV
    * Unit testing/CI set up. Basic unit tests for built-in outputs and history handling.
    * Test added for the fixed history bug.
    * Test added for legacy history conversion, which was kind of just implemented in the previous
      version, without testing.

3.0.2 (2018-08-30):
Multi-output bugfixes
* BREAKING CHANGES
    * upload_media had to go to make multi-output work in a reasonable way. I did this without
      thinking, which is why 2.2.0 wasn't marked as breaking. I realized I didn't have an easy way
      to restore it, so I'm just admitting this is breaking now.  Replace calls to upload_media
      followed by send_with_many_media just with send_with_many_media - it will do the upload_media
      appropriate for the given service.

* BUGFIXES
    * send_with_many_media in outputs has unnecessary (and broken?) use of splat operator
    * mastodon actually extremely misused splats
    * didn't pass up extra keys while adapting old birdsite records to new multirecords
    * mastodon output didn't pass media ids when sending correctly, meaning images never made it to
      toots

2.2.0 (2018-08-17):
Multi-output
* FEATURES
    * Support more than one site!!!
    * Rework output from skeleton so it doesn't assume only one site output.
    * Move birdsite to a new output separate from the main one.
    * Add support for Mastodon.
    * Change record formatting to multi-output site, convert old records in-place.

NOTE: entries beyond this point are backfilled. They should be relatively accurate, but won't be
exact.

2.1.0 (not actually tagged)
Error handling
* DEV/BUGFIXES
    * add framework for handling errors
    * call error handler for errors
    * will call out to dm_sos if it can't handle the error
    * add default handler for birdsite's duplicate tweet error, which bots hit a lot

2.0.5 (also 2.0.3-2.0.4) (2018-05-22):
Bugfixes
* DEV/BUGFIXES
    * pass over errors a few more times to make them store correctly and not crash the program
    * JSON-dump their dict when creating the TweetRecord
    * also do some type-checking when trying to assume things about error structure

2.0.2 (2018-04-18):
DREWTILITIES
* DEV/BUGFIX
    * we expect errors to be objects, so just pass the whole object in TweetRecords

2.0.2 (2018-04-18):
DREWTILITIES
* DEV/BUGFIX
    * we expect errors to be objects, so just pass the whole object in TweetRecords

2.0.1 (2018-01-02):
DREWTILITIES
* DEV/BUGFIX
    * errors can't be indexed, just stringify them when storing in TweetRecords

2.0.0 (2017-12-14):
DREWTILITIES
* DEV
    * use drewtilities package (same author) for random_line, log setup, rate limiting
    * remove those methods from this package
    * store log object on botskeleton object
    * store error records when errors happen
    * upload_media should also store history items
    * TweepErrors are a supertype of tweepy.RateLimitError, so we don't have to catch both
    * remove some unused variables

botskeleton/test/test_botskeleton.py
"""Tests for base botskeleton."""
import os
from shutil import copyfile
from typing import Generator

import pytest

import botskeleton

HERE = os.path.abspath(os.path.dirname(__file__))
JSON = os.path.join(HERE, "json")


def test_no_secrets_dir_fails(log: str) -> None:
    try:
        bs = botskeleton.BotSkeleton(log_filename=log)
        pytest.fail("Should throw exception for no secrets dir.")
    except botskeleton.BotSkeletonException as e:
        pass


def test_outputs_start_inactive(testdir: str, log: str) -> None:
    bs = botskeleton.BotSkeleton(secrets_dir=testdir, log_filename=log)

    # it would be funny to make this a mapped lambda
    for _, output in bs.outputs.items():
        assert not output["active"]


def test_load_null_history(testdir: str, log: str) -> None:
    name = "foobot"
    bs = botskeleton.BotSkeleton(bot_name=name, secrets_dir=testdir, log_filename=log)

    assert bs.history == []

    open(bs.history_filename, "w").close()
    bs.load_history()
    assert bs.history == []

    os.remove(bs.history_filename)
    os.remove(f"{bs.history_filename}.bak")


def test_load_existing_modern_history(testdir: str, testhist: str, log: str) -> None:
    bs = botskeleton.BotSkeleton(secrets_dir=testdir, history_filename=testhist, log_filename=log)
    bs.load_history()

    assert bs.history != []
    assert len(bs.history) == 2


def test_convert_legacy_history(testdir: str, legacyhist: str, hoistedhist: str, log: str) -> None:
    bs = botskeleton.BotSkeleton(secrets_dir=testdir, history_filename=legacyhist,
                                 log_filename=log)
    bs.load_history()

    assert bs.history != []

    mbs = botskeleton.BotSkeleton(secrets_dir=testdir, history_filename=hoistedhist,
                                  log_filename=log)
    mbs.load_history()

    assert len(bs.history) == len(mbs.history)
    for i, elem in enumerate(bs.history):
        melem = mbs.history[i]
        for key, item in elem.__dict__.items():
            assert str(item) == str(melem.__dict__[key])


# regression test for the history corruption snafu.
def test_repair_corrupted_history(testdir: str, corruptedhist: str, repairedcorruptedhist: str,
                                  log: str
                                  ) -> None:
    bs = botskeleton.BotSkeleton(secrets_dir=testdir, history_filename=corruptedhist,
                                 log_filename=log)
    bs.load_history()

    assert bs.history != []

    # compare against repairedhist. they should be identical.
    mbs = botskeleton.BotSkeleton(secrets_dir=testdir, history_filename=repairedcorruptedhist,
                                  log_filename=log)
    mbs.load_history()

    assert mbs.history != []

    assert len(bs.history) == len(mbs.history)
    for i, elem in enumerate(bs.history):
        melem = mbs.history[i]
        for key, item in elem.__dict__.items():
            assert str(item) == str(melem.__dict__[key])


def test_idempotency(testdir: str, testhist: str, log: str) -> None:
    bs = botskeleton.BotSkeleton(secrets_dir=testdir, history_filename=testhist, log_filename=log)
    bs.load_history()

    # maybe find a less hacky way to do this.
    for entry in bs.history:
        entry._version = "9999"
    bs.update_history()

    identical = True
    hist_source = os.path.join(JSON, "test_entries.json")
    with open(testhist, "r") as f1, open(hist_source, "r") as f2:
        for line1, line2 in zip(f1, f2):
            if line1 != line2:
                identical = False
                break

    if not identical:
        pytest.fail("Test history changed when it shouldn't have been.")


@pytest.fixture(scope="function")
def testhist(testdir: str) -> Generator[str, str, None]:
    hist_source = os.path.join(JSON, "test_entries.json")
    hist_file = os.path.join(testdir, "test.json")
    copyfile(hist_source, hist_file)
    yield hist_file
    os.remove(hist_file)


@pytest.fixture(scope="function")
def legacyhist(testdir: str) -> Generator[str, str, None]:
    hist_source = os.path.join(JSON, "legacy_entries.json")
    hist_file = os.path.join(testdir, "legacy.json")
    copyfile(hist_source, hist_file)
    yield hist_file
    os.remove(hist_file)


@pytest.fixture(scope="function")
def hoistedhist(testdir: str) -> Generator[str, str, None]:
    hist_source = os.path.join(JSON, "legacy_hoisted_entries.json")
    hist_file = os.path.join(testdir, "hoisted.json")
    copyfile(hist_source, hist_file)
    yield hist_file
    os.remove(hist_file)


@pytest.fixture(scope="function")
def corruptedhist(testdir: str) -> Generator[str, str, None]:
    hist_source = os.path.join(JSON, "corrupted_entries.json")
    hist_file = os.path.join(testdir, "corrupted.json")
    copyfile(hist_source, hist_file)
    yield hist_file
    os.remove(hist_file)


@pytest.fixture(scope="function")
def repairedcorruptedhist(testdir: str) -> Generator[str, str, None]:
    hist_source = os.path.join(JSON, "repaired_corrupted_entries.json")
    hist_file = os.path.join(testdir, "repaired.json")
    copyfile(hist_source, hist_file)
    yield hist_file
    os.remove(hist_file)


@pytest.fixture(scope="module")
def log(testdir: str) -> Generator[str, str, None]:
    log = os.path.join(testdir, "log")
    open(log, "a").close()
    yield log
    os.remove(log)


@pytest.fixture(scope="module")
def testdir() -> Generator[str, str, None]:
    directory = os.path.join(HERE, "testing_playground")
    os.mkdir(directory)
    yield directory
    os.rmdir(directory)


botskeleton/test/test_output_birdsite.py
import os
from shutil import copyfile
from typing import Any, Generator

import pytest

import botskeleton

HERE = os.path.abspath(os.path.dirname(__file__))

def test_birdsite_activates_correctly(testdir: str, credentials: str, log: str) -> None:
    bs = botskeleton.BotSkeleton(secrets_dir=testdir, log_filename=log)

    assert(bs.outputs["birdsite"]["active"])

def test_birdsite_activates_with_owner_id_correctly(testdir: str, credentials: str, log: str
                                                    ) -> None:
    # no owner set
    bs = botskeleton.BotSkeleton(secrets_dir=testdir, log_filename=log)

    assert(bs.outputs["birdsite"]["active"])
    birdsite_obj: Any = bs.outputs["birdsite"]["obj"]
    assert(birdsite_obj.owner_handle == "")

    TESTSTR = "foo"
    TESTFILE = os.path.join(credentials, "OWNER_HANDLE")
    with open(TESTFILE, "w") as f:
        f.write(TESTSTR)

    bs = botskeleton.BotSkeleton(secrets_dir=testdir, log_filename=log)

    assert(bs.outputs["birdsite"]["active"])
    birdsite_obj = bs.outputs["birdsite"]["obj"]
    assert(birdsite_obj.owner_handle == TESTSTR)

    os.remove(TESTFILE)


@pytest.fixture(scope="function")
def credentials(testdir: str) -> Generator[str, str, None]:
    credentials_birdsite = os.path.join(testdir, "credentials_birdsite")
    os.mkdir(credentials_birdsite)

    files = ["CONSUMER_KEY", "CONSUMER_SECRET", "ACCESS_SECRET", "ACCESS_TOKEN"]
    for file in files:
        open(os.path.join(credentials_birdsite, file), "a").close()

    yield credentials_birdsite

    for file in files:
        os.remove(os.path.join(credentials_birdsite, file))

    os.rmdir(credentials_birdsite)


@pytest.fixture(scope="module")
def log(testdir: str) -> Generator[str, str, None]:
    log = os.path.join(testdir, "log")
    open(log, "a").close()
    yield log
    os.remove(log)


@pytest.fixture(scope="module")
def testdir() -> Generator[str, str, None]:
    directory = os.path.join(HERE, "testing_playground")
    os.mkdir(directory)

    yield directory

    os.rmdir(directory)


botskeleton/test/test_output_mastodon.py
import os
from shutil import copyfile
from typing import Any, Generator

import pytest

import botskeleton

HERE = os.path.abspath(os.path.dirname(__file__))

def test_mastodon_activates_correctly(testdir: str, credentials: str, log: str) -> None:
    bs = botskeleton.BotSkeleton(secrets_dir=testdir, log_filename=log)

    assert(bs.outputs["mastodon"]["active"])

def test_mastodon_activates_with_instance_id_correctly(testdir: str, credentials: str, log: str
                                                       ) -> None:
    # no instance set
    bs = botskeleton.BotSkeleton(secrets_dir=testdir, log_filename=log)

    assert(bs.outputs["mastodon"]["active"])
    mastodon_obj: Any = bs.outputs["mastodon"]["obj"]
    assert(mastodon_obj.instance_base_url == "https://mastodon.social")

    TESTSTR = "https://butts.butts"
    TESTFILE = os.path.join(credentials, "INSTANCE_BASE_URL")
    with open(TESTFILE, "w") as f:
        f.write(TESTSTR)

    bs = botskeleton.BotSkeleton(secrets_dir=testdir, log_filename=log)

    assert(bs.outputs["mastodon"]["active"])
    mastodon_obj = bs.outputs["mastodon"]["obj"]
    assert(mastodon_obj.instance_base_url == TESTSTR)

    os.remove(TESTFILE)


@pytest.fixture(scope="function")
def credentials(testdir: str) -> Generator[str, str, None]:
    credentials_mastodon = os.path.join(testdir, "credentials_mastodon")
    os.mkdir(credentials_mastodon)

    files = ["ACCESS_TOKEN"]
    for file in files:
        open(os.path.join(credentials_mastodon, file), "a").close()

    yield credentials_mastodon

    for file in files:
        os.remove(os.path.join(credentials_mastodon, file))

    os.rmdir(credentials_mastodon)


@pytest.fixture(scope="module")
def log(testdir: str) -> Generator[str, str, None]:
    log = os.path.join(testdir, "log")
    open(log, "a").close()
    yield log
    os.remove(log)


@pytest.fixture(scope="module")
def testdir() -> Generator[str, str, None]:
    directory = os.path.join(HERE, "testing_playground")
    os.mkdir(directory)

    yield directory

    os.rmdir(directory)


botskeleton/outputs/output_utils.py
"""Stuff used by output classes."""
from datetime import datetime
from logging import Logger
from typing import Any, Callable, Dict, List

class OutputSkeleton:
    """Common stuff for output skeletons."""
    def __init__(
            self,
            *,
            secrets_dir: str,
            log: Logger,
            bot_name: str,
    ) -> None:
        self.log = log
        self.secrets_dir = secrets_dir

        self.bot_name = bot_name
        self.handled_errors: Dict[int, Any] = {}

        # Output skeletons must implement these.
        # mypy doesn't let us express a function taking only keyword arguments,
        # as best I can tell.
        self.cred_init: Callable[..., None]
        self.send: Callable[..., List[OutputRecord]]
        self.send_with_media: Callable[..., List[OutputRecord]]
        self.perform_batch_reply: Callable[..., List[OutputRecord]]

    def linfo(self, message: str) -> None:
        """Wrapped debug log with prefix key."""
        self.log.info(f"{self.bot_name}: {message}")

    def ldebug(self, message: str) -> None:
        """Wrapped debug log with prefix key."""
        self.log.debug(f"{self.bot_name}: {message}")

    def lerror(self, message: str) -> None:
        """Wrapped error log with prefix key."""
        self.log.error(f"{self.bot_name}: {message}")

class OutputRecord:
    """Record for an output occurrence."""
    def __init__(self) -> None:
        """Create tweet record object."""
        self._type = self.__class__.__name__
        self.timestamp = datetime.now().isoformat()

    def __str__(self) -> str:
        """Print object."""
        return str(self.__dict__)

    def __repr__(self) -> str:
        """repr object"""
        return str(self)

    def __eq__(self, other:Any) -> bool:
        """Overrides the default implementation"""
        if isinstance(other, OutputRecord):
            for key, value in self.__dict__.items():
                if value != other.__dict__[key]:
                    return False

                return True
        return False

    @classmethod
    def from_dict(cls, obj_dict: Dict[str, Any]) -> "OutputRecord":
        """Get object back from dict."""
        obj = cls()
        for key, item in obj_dict.items():
            obj.__dict__[key] = item

        return obj


botskeleton/outputs/output_birdsite.py
"""Skeleton code for sending to the bad bird site."""
import html
import json
from logging import Logger
from os import path
from typing import Any, Callable, Dict, List, Optional, Tuple

import tweepy

from .output_utils import OutputRecord, OutputSkeleton


class BirdsiteSkeleton(OutputSkeleton):
    def __init__(self) -> None:
        """Set up birdsite skeleton stuff."""
        self.name = "BIRDSITE"

        self.handled_errors = {
            187: self.default_duplicate_handler,
        }

    ## API implementation methods.
    def cred_init(
            self,
            *,
            secrets_dir: str,
            log: Logger,
            bot_name: str,
    ) -> None:
        """
        Initialize what requires credentials/secret files.

        :param secrets_dir: dir to expect credentials in and store logs/history in.
        :param log: logger to use for log output.
        :param bot_name: name of this bot,
            used for various kinds of labelling.
        :returns: none.
        """
        super().__init__(secrets_dir=secrets_dir, log=log, bot_name=bot_name)

        self.ldebug("Retrieving CONSUMER_KEY...")
        with open(path.join(self.secrets_dir, "CONSUMER_KEY")) as f:
            CONSUMER_KEY = f.read().strip()

        self.ldebug("Retrieving CONSUMER_SECRET...")
        with open(path.join(self.secrets_dir, "CONSUMER_SECRET")) as f:
            CONSUMER_SECRET = f.read().strip()

        self.ldebug("Retrieving ACCESS_TOKEN...")
        with open(path.join(self.secrets_dir, "ACCESS_TOKEN")) as f:
            ACCESS_TOKEN = f.read().strip()

        self.ldebug("Retrieving ACCESS_SECRET...")
        with open(path.join(self.secrets_dir, "ACCESS_SECRET")) as f:
            ACCESS_SECRET = f.read().strip()

        self.ldebug("Looking for OWNER_HANDLE...")
        owner_handle_path = path.join(self.secrets_dir, "OWNER_HANDLE")
        if path.isfile(owner_handle_path):
            with open(owner_handle_path) as f:
                self.owner_handle = f.read().strip()
        else:
            self.ldebug("Couldn't find OWNER_HANDLE, unable to DM...")
            self.owner_handle = ""

        self.auth = tweepy.OAuthHandler(CONSUMER_KEY, CONSUMER_SECRET)
        self.auth.set_access_token(ACCESS_TOKEN, ACCESS_SECRET)

        self.api = tweepy.API(self.auth)

    def send(
            self,
            *,
            text: str,
    ) -> List[OutputRecord]:
        """
        Send birdsite message.

        :param text: text to send in post.
        :returns: list of output records,
            each corresponding to either a single post,
            or an error.
        """
        try:
            status = self.api.update_status(text)
            self.ldebug(f"Status object from tweet: {status}.")
            return [TweetRecord(record_data={"tweet_id": status._json["id"], "text": text})]

        except tweepy.TweepError as e:
            return [self.handle_error(
                message=(f"Bot {self.bot_name} encountered an error when "
                 f"sending post {text} without media:\n{e}\n"),
                error=e)]

    def send_with_media(
            self,
            *,
            text: str,
            files: List[str],
            captions: List[str]=None
    ) -> List[OutputRecord]:
        """
        Upload media to birdsite,
        and send status and media,
        and captions if present.

        :param text: tweet text.
        :param files: list of files to upload with post.
        :param captions: list of captions to include as alt-text with files.
        :returns: list of output records,
            each corresponding to either a single post,
            or an error.
        """

        # upload media
        media_ids = None
        try:
            self.ldebug(f"Uploading files {files}.")
            media_ids = [self.api.media_upload(file).media_id_string for file in files]
        except tweepy.TweepError as e:
            return [self.handle_error(
                message=f"Bot {self.bot_name} encountered an error when uploading {files}:\n{e}\n",
                error=e)]

        # apply captions, if present
        self._handle_caption_upload(media_ids, captions)

        # send status
        try:
            status = self.api.update_status(status=text, media_ids=media_ids)
            self.ldebug(f"Status object from tweet: {status}.")
            return [TweetRecord(record_data={
                "tweet_id": status._json["id"],
                "text": text,
                "media_ids": media_ids,
                "captions": captions,
                "files": files
            })]

        except tweepy.TweepError as e:
            return [self.handle_error(
                message=(f"Bot {self.bot_name} encountered an error when "
                 f"sending post {text} with media ids {media_ids}:\n{e}\n"),
                error=e)]

    def perform_batch_reply(
            self,
            *,
            callback: Callable[..., str],
            lookback_limit: int,
            target_handle: str,
    ) -> List[OutputRecord]:
        """
        Performs batch reply on target account.
        Looks up the recent messages of the target user,
        applies the callback,
        and replies with
        what the callback generates.

        :param callback: a callback taking a message id,
            message contents,
            and optional extra keys,
            and returning a message string.
        :param target: the id of the target account.
        :param lookback_limit: a lookback limit of how many messages to consider.
        :returns: list of output records,
            each corresponding to either a single post,
            or an error.
        """
        self.log.info(f"Attempting to batch reply to birdsite user {target_handle}")

        if "@" in target_handle:
            base_target_handle = target_handle[1:]
        else:
            base_target_handle = target_handle

        records: List[OutputRecord] = []
        statuses = self.api.user_timeline(screen_name=base_target_handle, count=lookback_limit)
        self.log.debug(f"Retrieved {len(statuses)} statuses.")
        for i, status in enumerate(statuses):
            self.log.debug(f"Processing status {i} of {len(statuses)}")
            status_id = status.id

            # find possible replies we've made.
            # the 10 * lookback_limit is a guess,
            # might not be enough and I'm not sure we can guarantee it is.
            our_statuses = self.api.user_timeline(since_id=status_id,
                                                  count=lookback_limit * 10)
            in_reply_to_ids = list(map(lambda x: x.in_reply_to_status_id, our_statuses))

            if status_id not in in_reply_to_ids:
                # the twitter API and tweepy will attempt to give us the truncated text of the
                # message if we don't do this roundabout thing.
                encoded_status_text = self.api.get_status(status_id,
                                                  tweet_mode="extended")._json["full_text"]

                status_text = html.unescape(encoded_status_text)
                message = callback(message_id=status_id, message=status_text, extra_keys={})

                full_message = f"@{base_target_handle} {message}"
                self.log.info(f"Trying to reply with {message} to status {status_id} "
                              f"from {target_handle}.")
                try:
                    new_status = self.api.update_status(status=full_message,
                                                        in_reply_to_status_id=status_id)

                    records.append(TweetRecord(record_data={
                        "tweet_id": new_status.id,
                        "in_reply_to": f"@{base_target_handle}",
                        "in_reply_to_id": status_id,
                        "text": full_message,
                    }))

                except tweepy.TweepError as e:
                    records.append(self.handle_error(
                        message=(f"Bot {self.bot_name} encountered an error when "
                         f"trying to reply to {status_id} with {message}:\n{e}\n"),
                        error=e))
            else:
                self.log.info(f"Not replying to status {status_id} from {target_handle} "
                              f"- we already replied.")

        return records

    ## Helpful methods for this output.
    def send_dm_sos(self, message: str) -> None:
        """
        Send DM to owner if something happens.

        :param message: message to send to owner.
        :returns: None.
        """
        if self.owner_handle:
            try:
                # twitter changed the DM API and tweepy (as of 2019-03-08)
                # has not adapted.
                # fixing with
                # https://github.com/tweepy/tweepy/issues/1081#issuecomment-423486837
                owner_id = self.api.get_user(screen_name=self.owner_handle).id
                event = {
                    "event": {
                        "type": "message_create",
                        "message_create": {
                            "target": {
                                "recipient_id": f"{owner_id}",
                            },
                            "message_data": {
                                "text": message
                            }
                        }
                    }
                }

                self._send_direct_message_new(event)

            except tweepy.TweepError as de:
                self.lerror(f"Error trying to send DM about error!: {de}")

        else:
            self.lerror("Can't send DM SOS, no owner handle.")

    def handle_error(
            self,
            *,
            message: str,
            error: tweepy.TweepError,
    ) -> OutputRecord:
        """
        Handle error while trying to do something.

        :param message: message to send in DM regarding error.
        :param e: tweepy error object.
        :returns: OutputRecord containing an error.
        """
        self.lerror(f"Got an error! {error}")

        # Handle errors if we know how.
        try:
            code = error[0]["code"]
            if code in self.handled_errors:
                self.handled_errors[code]
            else:
                self.send_dm_sos(message)

        except Exception:
            self.send_dm_sos(message)

        return TweetRecord(error=error)

    def default_duplicate_handler(self) -> None:
        """Default handler for duplicate status error."""
        self.linfo("Duplicate handler: who cares about duplicate statuses.")
        return

    def set_duplicate_handler(self, duplicate_handler: Callable[..., None]) -> None:
        self.handled_errors[187] = duplicate_handler

    def _handle_caption_upload(self, media_ids: List[str], captions: Optional[List[str]]) -> None:
        """
        Handle uploading all captions.

        :param media_ids: media ids of uploads to attach captions to.
        :param captions: captions to be attached to those media ids.
        :returns: None.
        """
        if captions is None:
            return

        if len(media_ids) > len(captions):
            captions.extend([""] * (len(media_ids) - len(captions)))

        for i, media_id in enumerate(media_ids):
            caption = captions[i]
            self._upload_caption(media_id=media_id, caption=caption)

    # taken from https://github.com/tweepy/tweepy/issues/716#issuecomment-398844271
    def _upload_caption(self, media_id: str, caption: str) -> Any:
        post_data = {
            "media_id": media_id,
            "alt_text": {
                "text": caption,
            },
        }

        metadata_path = "/media/metadata/create.json"

        return tweepy.binder.bind_api(
            api=self.api,
            path=metadata_path,
            method="POST",
            allowed_param=[],
            require_auth=True,
            upload_api=True,
        )(post_data=json.dumps(post_data))

    # taken from
    # https://github.com/do-n-khanh/tweepy/commit/79772c976c64830149095f087c16c181912466ba#diff-ea5dd38a4efd9ff36c96e04ab0597cfb
    def _send_direct_message_new(self, messageobject: Dict[str, Dict]) -> Any:
        """
        :reference: https://developer.twitter.com/en/docs/direct-messages/sending-and-receiving/api-reference/new-event.html
        """
        headers, post_data = _buildmessageobject(messageobject)
        newdm_path = "/direct_messages/events/new.json"

        return tweepy.binder.bind_api(
            api=self.api,
            path=newdm_path,
            method="POST",
            require_auth=True,
        )(post_data=post_data, headers=headers)

# taken from
# https://github.com/do-n-khanh/tweepy/commit/79772c976c64830149095f087c16c181912466ba#diff-ea5dd38a4efd9ff36c96e04ab0597cfb
def _buildmessageobject(messageobject: Dict[str, Dict]) -> Tuple[Dict[str, str], str]:
    body = json.dumps(messageobject)
    # build headers
    headers = {
        "Content-Type": "application/json",
        "Content-Length": str(len(body))
    }

    return headers, body

class TweetRecord(OutputRecord):
    def __init__(
            self,
            *,
            record_data: Dict[str, Any]={},
            error: tweepy.TweepError=None,
    ) -> None:
        """
        Create tweet record object.

        :param record_data: data to use to generate a TweetRecord.
        :param error: error encountered while posting,
            to generate a record with.
        """
        super().__init__()
        self._type = self.__class__.__name__
        self.tweet_id = record_data.get("tweet_id", None)
        self.id = self.tweet_id
        self.text = record_data.get("text", None)
        self.files = record_data.get("files", [])
        self.media_ids = record_data.get("media_ids", [])
        self.captions = record_data.get("captions", [])
        self.timestamp = record_data.get("timestamp", None)
        self.in_reply_to = record_data.get("in_reply_to", None)
        self.in_reply_to_id = record_data.get("in_reply_to_id", None)

        if error is not None:
            # So Python doesn't get upset when we try to json-dump the record later.
            try:
                self.error = json.dumps(error.__dict__)
            except TypeError:
                # fuck you
                self.error = str(error)

            try:
                if isinstance(error.message, str):
                    self.error_message = error.message
                elif isinstance(error.message, list):
                    self.error_code = error.message[0]["code"]
                    self.error_message = error.message[0]["message"]
            except AttributeError:
                # fine, I didn't want it anyways.
                pass
